### PR TITLE
Resolve realpath

### PR DESCRIPTION
**Pull request by dishuostec**
Origin pull request https://github.com/parcel-bundler/parcel/pull/4939

I fixed all test https://github.com/parcel-bundler/parcel/pull/4939#issuecomment-684463349

### CODE REVIEW

**Comment by aminya** [comment]
Wow! This is awesome. Excited to see this in action.

**Comment by jasonswearingen** [comment]
> Should we revive the [pnpm PR](https://github.com/parcel-bundler/parcel/pull/4920) as well after this is merged?

Please do, I have been tracking this issue for that very reason.

**Comment by dishuostec** [comment]
> Wow! This is awesome. Excited to see this in action. Should we revive the [pnpm PR](https://github.com/parcel-bundler/parcel/pull/4920) as well after this is merged?

I have tested pnpm with this pr, it works fine and no more duplicate depenndency.

**Comment by aminya** [comment]
I want to request the maintainers to review this if they can and let us know if there is anything we can do to get this through. I'm afraid that other pull requests introduce conflicts with this, like how it happened for the old PR. 😄 

cc: @mischnic @DeMoorJasper @devongovett etc

File: packages/utils/node-resolver-core/src/NodeResolver.js [hunk 1]
```diff
@@ -429,11 +430,16 @@ export default class NodeResolver {
         let moduleDir = path.join(dir, 'node_modules', moduleName);
         let stats = await this.fs.stat(moduleDir);
         if (stats.isDirectory()) {
+          const realModuleDir = this.fs.realpathSync(moduleDir);
```
Comment (devongovett) [inline packages/utils/node-resolver-core/src/NodeResolver.js:433]: Can we avoid sync fs calls and use the async version instead?

File: packages/utils/node-resolver-core/src/NodeResolver.js [hunk 2]
```diff
@@ -598,8 +611,7 @@ export default class NodeResolver {
     // If the package has a `source` field, check if it is behind a symlink.
     // If so, we treat the module as source code rather than a pre-compiled module.
     if (pkg.source) {
-      let realpath = await this.fs.realpath(file);
-      if (realpath === file) {
+      if (!isSymlink) {
         delete pkg.source;
```
Comment (devongovett) [inline packages/utils/node-resolver-core/src/NodeResolver.js:615]: I'm curious: what fails if you get rid of this logic. I'm wondering if we could get rid of it entirely.

**Review by devongovett — CHANGES_REQUESTED** [review]

File: packages/utils/node-resolver-core/src/NodeResolver.js (hunk 1 above)
Comment (aminya) [reply packages/utils/node-resolver-core/src/NodeResolver.js:433]:
```suggestion
          const realModuleDir = await this.fs.realpath(moduleDir);
```

File: packages/utils/node-resolver-core/src/NodeResolver.js (hunk 2 above)
Comment (aminya) [reply packages/utils/node-resolver-core/src/NodeResolver.js:615]: What about postponing micro-optimizations for follow up PRs?

**Comment by aminya** [comment]
Anything else remaining here? What's the procedure for merging something into Parcel? It's a little unclear how the PRs work in Parcel considering that this (or its related PR) has been open for a couple of months now.

File: packages/utils/node-resolver-core/src/NodeResolver.js (hunk 2 above)
Comment (mischnic) [reply packages/utils/node-resolver-core/src/NodeResolver.js:615]: This test fails: https://github.com/parcel-bundler/parcel/blob/146cffb6896d1f2bcc3f7f4426440b09b61142b7/packages/utils/node-resolver-core/test/resolver.js#L701-L717

Comment (aminya) [reply packages/utils/node-resolver-core/src/NodeResolver.js:615]: Someone should sit and read all of these changes to figure out what has changed where that this test does not pass anymore
https://github.com/dishuostec/parcel/compare/resolve-realpath...parcel-bundler:v2

**Comment by jasonswearingen** [comment]
@devongovett please consider accepting this PR.  It or it's predecessor have been around for 5 months and dishuostec got it mergeable twice.

**Comment by aminya** [comment]
I have made a fork of Parcel that includes this PR and #4920. However, I am not familiar with the registration procedure for this repository because it uses Lerna, Yarn, Gulp, etc. If anyone is familiar with the process, let me know in this issue:
https://github.com/aminya/parcel/issues/4

**Comment by aminya** [comment]
I merged V2 into this branch, and ran the tests. Everything works without any issues.

I only had to increase this time out to `30000`
https://github.com/parcel-bundler/parcel/blob/a4cf3e958ea1136b4829eb93615fba5328213471/packages/core/core/test/AssetGraphBuilder.test.js#L11

**Comment by aminya** [comment]
@DeMoorJasper Why is bundle undefined in parcel-benchmark-action? This seems strange.
https://github.com/parcel-bundler/parcel-benchmark-action/blob/75a86fd8b9815657df9a8f46a9ab762016ed2838/src/utils/compare-benchmarks.ts#L46

**Review by devongovett — COMMENTED** [review]
I apologize for the slow response to this PR. I am also making some changes to the resolver in #5481, so I will update this and make a new PR with those changes merged in. Thanks again for your contribution!

**Comment by devongovett** [comment]
See https://github.com/parcel-bundler/parcel/pull/5508

**Pull request closed without merging**